Lib/annotationlib.py

```diff
@@ -639,28 +639,38 @@ def get_annotations(
     if eval_str and format != Format.VALUE:
         raise ValueError("eval_str=True is only supported with format=Format.VALUE")
 
-    # For VALUE format, we look at __annotations__ directly.
-    if format != Format.VALUE:
-        annotate = get_annotate_function(obj)
-        if annotate is not None:
-            ann = call_annotate_function(annotate, format, owner=obj)
-            if not isinstance(ann, dict):
-                raise ValueError(f"{obj!r}.__annotate__ returned a non-dict")
-            return dict(ann)
-
-    if isinstance(obj, type):
-        try:
-            ann = _BASE_GET_ANNOTATIONS(obj)
-        except AttributeError:
-            # For static types, the descriptor raises AttributeError.
-            return {}
-    else:
-        ann = getattr(obj, "__annotations__", None)
-        if ann is None:
-            return {}
-
-    if not isinstance(ann, dict):
-        raise ValueError(f"{obj!r}.__annotations__ is neither a dict nor None")
+    match format:
+        case Format.VALUE:
+            # For VALUE, we only look at __annotations__
+            ann = _get_dunder_annotations(obj)
+        case Format.FORWARDREF:
+            # For FORWARDREF, we use __annotations__ if it exists
+            try:
+                ann = _get_dunder_annotations(obj)
+            except NameError:
```

Comment:
Hmm, I didn't anticipate that we'd do this by catching `NameError`, I thought we'd actually check the object's `__dict__` to see if it has a cached `__annotations__` or not. But I can see multiple pros and cons in both directions here:

* Catching `NameError` pokes at internals less / assumes less about the behavior of the `__annotations__` descriptor.
* Catching `NameError` means that the first call asking for `FORWARDREF` when it is actually needed will be significantly slower, since it will run `__annotate__` twice, raising and catching a `NameError` the first time.
* Catching `NameError` means that repeated requests for `FORWARDREF` when it _isn't_ required (`VALUE` would also have worked and returned the same thing) will actually cache `__annotations__`. This is potentially a significant positive.

Having written it all out, I think catching `NameError` is probably the better tradeoff :) But I'll go ahead and post the comment, in case it prompts other considerations that I missed.

Reply:
For function objects, there isn't a way to check whether `__annotations__` already exists, because the annotations are not in the object dict. So unless we make bigger changes to how the runtime objects work, I don't see an alternative.